File: apps/api/src/my_bot_api/auth/security.py
import hashlib
import hmac
import secrets


def generate_otp_code() -> str:
    return f"{secrets.randbelow(1_000_000):06d}"


def generate_opaque_token(size: int = 32) -> str:
    return secrets.token_urlsafe(size)


def hash_otp(*, challenge_id: str, code: str, pepper: str) -> str:
    return hmac.new(
        pepper.encode(),
        f"{challenge_id}:{code}".encode(),
        hashlib.sha256,
    ).hexdigest()


def keyed_identifier(value: str, *, pepper: str) -> str:
    return hmac.new(pepper.encode(), value.encode(), hashlib.sha256).hexdigest()


def hash_session_token(token: str) -> bytes:
    return hashlib.sha256(token.encode()).digest()

File: apps/api/src/security.ts
import { createHash, createHmac, randomBytes, randomInt, timingSafeEqual } from 'node:crypto'

export const generateOtpCode = () => randomInt(0, 1_000_000).toString().padStart(6, '0')

export const generateOpaqueToken = (size = 32) => randomBytes(size).toString('base64url')

export const hashOtp = (challengeId: string, code: string, pepper: string) =>
  createHmac('sha256', pepper).update(`${challengeId}:${code}`).digest('hex')

export const keyedIdentifier = (value: string, pepper: string) =>
  createHmac('sha256', pepper).update(value).digest('hex')

export const hashSessionToken = (token: string) => createHash('sha256').update(token).digest()

export const signValue = (value: string, secret: string) =>
  `${value}.${createHmac('sha256', secret).update(value).digest('base64url')}`

export const safeEqual = (first: string, second: string) => {
  const firstBuffer = Buffer.from(first)
  const secondBuffer = Buffer.from(second)
  return firstBuffer.length === secondBuffer.length && timingSafeEqual(firstBuffer, secondBuffer)
}

export const verifySignedValue = (signedValue: string | undefined, secret: string) => {
  if (!signedValue) return undefined
  const separator = signedValue.lastIndexOf('.')
  if (separator <= 0) return undefined
  const value = signedValue.slice(0, separator)
  return safeEqual(signedValue.slice(separator + 1), signValue(value, secret).slice(separator + 1))
    ? value
    : undefined
}
